retrieval/discovery/ec2.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2015 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package discovery
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/log"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	ec2Label = model.MetaLabelPrefix + "ec2_"
+)
+
+// EC2Discovery periodically performs EC2-SD requests. It implements
+// the TargetProvider interface.
+type EC2Discovery struct {
+	aws            *aws.Config
+	tags           model.LabelSet
+	regexTags      map[string]*config.Regexp
+	done           chan struct{}
+	interval       time.Duration
+	port           int
+	instanceFilter []*ec2.Filter
+}
+
+// NewEC2Discovery returns a new EC2Discovery which periodically refreshes its targets.
+func NewEC2Discovery(conf *config.EC2SDConfig) *EC2Discovery {
+	creds := credentials.NewStaticCredentials(conf.AccessKey, conf.SecretKey, "")
+	if conf.AccessKey == "" && conf.SecretKey == "" {
+		creds = credentials.NewEnvCredentials()
+	}
+	tags := make(model.LabelSet)
+	filter := &ec2.Filter{
+		Name: aws.String("tag-key"),
+	}
+	for k, v := range conf.Tags {
+		tags[model.LabelName(k)] = model.LabelValue(v.String())
+		filter.Values = append(filter.Values, &k)
+	}
+	return &EC2Discovery{
+		aws: &aws.Config{
+			Region:      &conf.Region,
+			Credentials: creds,
+		},
+		tags:      tags,
+		regexTags: conf.Tags,
+		done:      make(chan struct{}),
+		interval:  time.Duration(conf.RefreshInterval),
+		port:      conf.Port,
+		instanceFilter: []*ec2.Filter{
+			filter,
+		},
+	}
+}
+
+// Run implements the TargetProvider interface.
+func (ed *EC2Discovery) Run(ch chan<- *config.TargetGroup, done <-chan struct{}) {
+	defer close(ch)
+
+	ticker := time.NewTicker(ed.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	if err := ed.refresh(ch); err != nil {
+		log.Error(err)
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := ed.refresh(ch); err != nil {
+				log.Error(err)
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+// Sources implements the TargetProvider interface.
+func (ed *EC2Discovery) Sources() []string {
+	return []string{*ed.aws.Region}
```

Comment:
You need to return all the sources here too, or targets will blip.

@fabxc This is roughly the reason I'm in favour of a full dump from Sources() rather than just the source names.